src/analysis/processing/qgsalgorithmlinedensity.cpp

```diff
@@ -0,0 +1,216 @@
+
+  mExtent = mSource->sourceExtent();
+  mCrs = mSource->sourceCrs();
+  mDa = QgsDistanceArea();
+
+  if ( context.project() )
+  {
+    mDa.setEllipsoid( context.project()->ellipsoid() );
+  }
+  mDa.setSourceCrs( mCrs, context.transformContext() );
+
+
+  //get cell midpoint from top left cell
+  QgsPoint firstCellMidpoint = QgsPoint( mExtent.xMinimum() + ( mPixelSize / 2 ), mExtent.yMaximum() - ( mPixelSize / 2 ) );
+  QgsCircle searchCircle = QgsCircle( firstCellMidpoint, mSearchRadius );
+  mSearchGeometry = QgsGeometry( searchCircle.toPolygon() );
+
+  return true;
+}
+
+QVariantMap QgsLineDensityAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  mFeatureWeights = QHash<QgsFeatureId, double>();
+  mIndex = QgsSpatialIndex( QgsSpatialIndex::FlagStoreFeatureGeometries );
+
+  QgsFeatureIterator fit = mSource->getFeatures();
+  QgsFeature f;
+
+  while ( fit.nextFeature( f ) )
+  {
+    mIndex.addFeature( f, QgsFeatureSink::FastInsert );
+
+    //default weight of lines is set to 1 if no field provided
+    double analysisWeight = 1;
+    if ( !mWeightField.isEmpty() )
+    {
+      analysisWeight = f.attribute( mWeightField ).toDouble();
+    }
+    mFeatureWeights.insert( f.id(), analysisWeight );
+  }
+
+
+  const QString outputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT" ), context );
+  QFileInfo fi( outputFile );
+  const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+  int rows = std::max( std::ceil( mExtent.height() / mPixelSize ), 1.0 );
+  int cols = std::max( std::ceil( mExtent.width() / mPixelSize ), 1.0 );
+
+  //build new raster extent based on number of columns and cellsize
+  //this prevents output cellsize being calculated too small
+  QgsRectangle rasterExtent = QgsRectangle( mExtent.xMinimum(), mExtent.yMaximum() - ( rows * mPixelSize ), mExtent.xMinimum() + ( cols * mPixelSize ), mExtent.yMaximum() );
+
+  QgsRasterFileWriter writer = QgsRasterFileWriter( outputFile );
+  writer.setOutputProviderKey( QStringLiteral( "gdal" ) );
+  writer.setOutputFormat( outputFormat );
+  std::unique_ptr<QgsRasterDataProvider > provider( writer.createOneBandRaster( Qgis::Float32, cols, rows, rasterExtent, mCrs ) );
+  if ( !provider )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output: %1" ).arg( outputFile ) );
+  if ( !provider->isValid() )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output %1: %2" ).arg( outputFile, provider->error().message( QgsErrorMessage::Text ) ) );
+
+  provider->setNoDataValue( 1, -9999 );
+
+  qgssize totalCellcnt = static_cast<qgssize>( rows ) * cols;
+  int cellcnt = 0;
+
+  for ( int row = 0; row < rows; row++ )
+  {
+    std::unique_ptr< QgsRasterBlock > rasterDataLine = qgis::make_unique< QgsRasterBlock >( Qgis::Float32, cols, 1 );
```

Comment:
You could optimise things further here and move rasterDataLine up out of the row loop, so that it's only created once and reused for every row. That'll avoid allocated memory after every row.